### ISSUE TITLE
False positive `reportOverlappingOverloads` in `Protocol` methods

### ISSUE THREAD

**Issue by jorenham** [microsoft/pyright#9662]
Consider the following structural types, that are applicable to e.g. `datetime.timedelta` and `datetime.datetime`:

```py
from datetime import datetime, timedelta
from typing import Any, Protocol, Self, overload

class Intra(Protocol):
    def __add__(self, d: Self, /) -> Self: ...
    def __sub__(self, d: Self, /) -> Self: ...

class Inter[D: Intra](Protocol):
    def __add__(self, d: D, /) -> Self: ...

    @overload
    def __sub__(self, d: D, /) -> Self: ...
    @overload
    def __sub__(self, x: Self, /) -> D: ...  # E: reportOverlappingOverload
```

```
Overload 2 for "__sub__" will never be used because its parameters overlap overload 1  (reportOverlappingOverload)
```

But this contradicts the following outcome:

```py
def sub_other[D: Intra](x: Inter[D], dx: D, /) -> Inter[D]:
    return x - dx

def sub_self[D: Intra](x: Inter[D], /) -> D:
    return x - x

d0: Intra = timedelta()
t0: Inter[timedelta] = datetime.fromtimestamp(0)

reveal_type(sub_other(t0, d0))  # Inter[timedelta]
reveal_type(sub_self(t0))  # timedelta
```

```
Type of "sub_other(t0, d0)" is "Inter[timedelta]"
Type of "sub_self(t0)" is "timedelta"
```

https://pyright-play.net/?strict=true&code=GYJw9gtgBAJghgFwKYIJYSVdAHMIGyIrpIA0UaGMSANgnAFCiQUCe2qAdgOZYS74oABXAIwAYzA1yAZVrByYAG5IQNMHBgMG4mnADO%2BqAElOCEHAAUIsGMk0AlAC4GUN7CTAoAfW%2BaYvpb68uQwTlByNApQAPQOUAC0AHwR8uEAdJmu7tRevvoArgBGgcFRoeGR0XGJKVUZWTp6hiZmqgDaACLhpuZwALrWohJSztluuT5%2BMAHeQSGw4Z3kNcmpUQ3p2u5QAALKquqa4x553oUlc2XRYVDLsfFr9VCZWzv7KmoaWjuT%2BcWlBYAD0qC1WKW6L0ybgAxFAAKLhEBIAQIADynz02A4PAxh2%2B2hhRO0kwu3lsAAsOpDehZBiDWsgQF1%2BqEGfdwYzqf0XDtkQgCiBOFAgYlYECSZ4oGTrl0emY6ZYGb1uStHhDee5%2BYLhaKEiLCcSGDAAAzyvpQAC8FBI1DoVgcDAQZq5zMoSDt9H6VsIyHd6WYEHd%2Bno-EsJsdDGRKjgNG8CHYSCCAMpqksztCEfiUDhKrdttoXqjSBjcYT2CTMvk6azsJtVELjCAA

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.392.
